File: Tests/Conformance/ServiceNetworkFacadePrototype05826.c
#include <r4os/r4os.h>
#include <string.h>

enum ReplyMode { MODE_OK, MODE_WOULD_BLOCK, MODE_TIMEOUT, MODE_RESET, MODE_PEER_CLOSE, MODE_NO_SERVICE };
static enum ReplyMode reply_mode;
static uint32_t open_handles;
static int endpoint_registered;
static int endpoint_ready;

static int32_t fake_open(const uint8_t *name, R4ServiceInfo *out) {
    if (reply_mode == MODE_NO_SERVICE || strcmp((const char *)name, "STOPPED") == 0) return R4OS_SERVICE_API_RESULT_NOT_RUNNING;
    *out = (R4ServiceInfo){0};
    out->handle = strcmp((const char *)name, "DNSSVC") == 0 ? 101u : (strcmp((const char *)name, "TCPSVC") == 0 ? 102u : (strcmp((const char *)name, "UDPSVC") == 0 ? 103u : 104u));
    ++open_handles;
    return R4OS_SERVICE_API_RESULT_OK;
}

static int32_t fake_close(uint32_t handle) {
    if (handle < 101u || handle > 104u || open_handles == 0u) return R4OS_SERVICE_API_RESULT_BAD_HANDLE;
    --open_handles;
    return R4OS_SERVICE_API_RESULT_OK;
}

static uint32_t lifecycle(void) {
    if (reply_mode == MODE_WOULD_BLOCK) return R4OS_NET_SERVICE_SOCKET_LIFECYCLE_WOULD_BLOCK;
    if (reply_mode == MODE_TIMEOUT) return R4OS_NET_SERVICE_SOCKET_LIFECYCLE_TIMEOUT;
    if (reply_mode == MODE_RESET) return R4OS_NET_SERVICE_SOCKET_LIFECYCLE_RESET;
    if (reply_mode == MODE_PEER_CLOSE) return R4OS_NET_SERVICE_SOCKET_LIFECYCLE_PEER_GONE;
    return R4OS_NET_SERVICE_SOCKET_LIFECYCLE_ACTIVE;
}

static uint32_t service_status(void) {
    if (reply_mode == MODE_WOULD_BLOCK) return R4OS_NET_SERVICE_STATUS_WOULD_BLOCK;
    if (reply_mode == MODE_TIMEOUT) return R4OS_NET_SERVICE_STATUS_TIMEOUT;
    return R4OS_NET_SERVICE_STATUS_OK;
}

static int32_t fake_call(uint32_t handle, uint16_t op, const uint8_t *request, uint32_t request_len, R4ServiceMessageHeader *header, uint8_t *response, uint32_t capacity, uint64_t timeout) {
    (void)timeout;
    *header = (R4ServiceMessageHeader){0}; header->magic = R4OS_SERVICE_API_MAGIC; header->version = R4OS_SERVICE_API_VERSION; header->op = op; header->status = R4OS_SERVICE_API_RESULT_OK;
    if (handle == 104u) {
        if (op == 7u) { if (request_len > capacity) return R4OS_SERVICE_API_RESULT_BUFFER_TOO_SMALL; memcpy(response, request, request_len); return (int32_t)request_len; }
        header->status = R4OS_SERVICE_API_RESULT_INVALID; return 0;
    }
    if (handle == 101u) {
        R4NetServiceDnsResult result = {0}; result.magic = R4OS_NET_SERVICE_DNS_RESULT_MAGIC; result.version = R4OS_NET_SERVICE_DNS_RESULT_VERSION; result.result = R4OS_DNS_RESULT_OK; result.flags = R4OS_NET_SERVICE_DNS_FLAG_OK; result.answer[0] = 10u; result.answer[2] = 2u; result.answer[3] = 15u;
        if (reply_mode == MODE_TIMEOUT) result.result = R4OS_DNS_RESULT_TIMEOUT; else if (reply_mode == MODE_WOULD_BLOCK) result.result = R4OS_DNS_RESULT_NXDOMAIN;
        if (capacity < sizeof(result)) return R4OS_SERVICE_API_RESULT_BUFFER_TOO_SMALL; memcpy(response, &result, sizeof(result)); return (int32_t)sizeof(result);
    }
    if (handle == 102u) {
        R4NetServiceTcpResult result = {0}; result.magic = R4OS_NET_SERVICE_TCP_RESULT_MAGIC; result.version = R4OS_NET_SERVICE_TCP_RESULT_VERSION; result.action = op; result.flags = R4OS_NET_SERVICE_TCP_FLAG_OK | R4OS_NET_SERVICE_TCP_FLAG_HANDLE_VALID; result.handle = 501u; result.remote_ip[0] = 10u; result.remote_ip[2] = 2u; result.remote_ip[3] = 2u; result.remote_port = 1234u; result.lifecycle_cause = lifecycle(); result.service_status = service_status();
        const char *payload = (op == R4OS_NET_SERVICE_OP_TCP_READ_RESULT && reply_mode == MODE_OK) ? "OK" : ""; uint32_t payload_len = (uint32_t)strlen(payload);
        result.bytes = op == R4OS_NET_SERVICE_OP_TCP_WRITE_RESULT ? request_len - 4u : payload_len;
        if (op == R4OS_NET_SERVICE_OP_TCP_CLOSE_RESULT || op == R4OS_NET_SERVICE_OP_TCP_CLOSE_LISTEN_RESULT) result.lifecycle_cause = R4OS_NET_SERVICE_SOCKET_LIFECYCLE_LOCAL_CLOSE;
        if (sizeof(result) + payload_len > capacity) return R4OS_SERVICE_API_RESULT_BUFFER_TOO_SMALL; memcpy(response, &result, sizeof(result)); memcpy(response + sizeof(result), payload, payload_len); return (int32_t)(sizeof(result) + payload_len);
    }
    if (handle == 103u) {
        R4NetServiceUdpResult result = {0}; result.magic = R4OS_NET_SERVICE_UDP_RESULT_MAGIC; result.version = R4OS_NET_SERVICE_UDP_RESULT_VERSION; result.action = op; result.flags = R4OS_NET_SERVICE_UDP_FLAG_OK | R4OS_NET_SERVICE_UDP_FLAG_HANDLE_VALID; result.handle = 601u; result.source_ip[0] = 10u; result.source_ip[2] = 2u; result.source_ip[3] = 3u; result.source_port = 7000u; result.dest_port = 8000u; result.lifecycle_cause = lifecycle(); result.service_status = service_status();
        const char *payload = (op == R4OS_NET_SERVICE_OP_UDP_RECV_RESULT && reply_mode == MODE_OK) ? "UDP" : ""; uint32_t payload_len = (uint32_t)strlen(payload);
        result.bytes = op == R4OS_NET_SERVICE_OP_UDP_SENDTO_RESULT ? request_len - 10u : payload_len;
        if (op == R4OS_NET_SERVICE_OP_UDP_CLOSE_RESULT) result.lifecycle_cause = R4OS_NET_SERVICE_SOCKET_LIFECYCLE_LOCAL_CLOSE;
        if (sizeof(result) + payload_len > capacity) return R4OS_SERVICE_API_RESULT_BUFFER_TOO_SMALL; memcpy(response, &result, sizeof(result)); memcpy(response + sizeof(result), payload, payload_len); return (int32_t)(sizeof(result) + payload_len);
    }
    return R4OS_SERVICE_API_RESULT_BAD_HANDLE;
}

static int32_t fake_register(const uint8_t *name, uint32_t flags, R4ServiceInfo *out) { (void)name; (void)flags; if (endpoint_registered) return R4OS_SERVICE_API_RESULT_INVALID; endpoint_registered = 1; *out = (R4ServiceInfo){0}; out->handle = 201u; return 0; }
static int32_t fake_unregister(uint32_t handle) { if (!endpoint_registered || handle != 201u) return R4OS_SERVICE_API_RESULT_BAD_HANDLE; endpoint_registered = 0; return 0; }
static int32_t fake_wait(uint32_t handle, uint64_t timeout) { (void)timeout; return endpoint_registered && handle == 201u ? (endpoint_ready ? 1 : 0) : R4OS_SERVICE_API_RESULT_BAD_HANDLE; }
static int32_t fake_recv(uint32_t handle, R4ServiceMessageHeader *header, uint8_t *out, uint32_t capacity) { if (!endpoint_registered || handle != 201u) return R4OS_SERVICE_API_RESULT_BAD_HANDLE; if (!endpoint_ready) { *header = (R4ServiceMessageHeader){0}; return 0; } if (capacity < sizeof(uint32_t)) return R4OS_SERVICE_API_RESULT_BUFFER_TOO_SMALL; *header = (R4ServiceMessageHeader){0}; header->magic = R4OS_SERVICE_API_MAGIC; header->request_id = 88u; *(uint32_t *)(void *)out = 42u; endpoint_ready = 0; return (int32_t)sizeof(uint32_t); }
static int32_t fake_reply(uint32_t handle, uint32_t request_id, int32_t status, const uint8_t *payload, uint32_t payload_len) { (void)payload; (void)payload_len; return handle == 201u && request_id == 88u && status == 0 ? 0 : R4OS_SERVICE_API_RESULT_INVALID; }
static int32_t fake_tcp_marker(uint8_t a, uint8_t b, uint8_t c, uint8_t d, uint16_t port) { (void)a; (void)b; (void)c; (void)d; (void)port; return -1; }

static R4Timeout poll_timeout(void) { R4Timeout value = {0}; value.kind = R4OS_TIMEOUT_KIND_POLL; return value; }
static R4Timeout forever_timeout(void) { R4Timeout value = {0}; value.kind = R4OS_TIMEOUT_KIND_FOREVER; return value; }

static void init_app(R4App *app, R4XStartR4Sys *sys, R4XStartR4Net *net) {
    *sys = (R4XStartR4Sys){0}; sys->service_open = (uintptr_t)&fake_open; sys->service_close = (uintptr_t)&fake_close; sys->service_call = (uintptr_t)&fake_call; sys->service_endpoint_register = (uintptr_t)&fake_register; sys->service_endpoint_unregister = (uintptr_t)&fake_unregister; sys->service_endpoint_wait = (uintptr_t)&fake_wait; sys->service_endpoint_recv = (uintptr_t)&fake_recv; sys->service_endpoint_reply = (uintptr_t)&fake_reply;
    *net = (R4XStartR4Net){0}; net->tcp_connect = (uintptr_t)&fake_tcp_marker;
    *app = (R4App){0}; app->system.table = sys; app->network.table = net;
}

int main(void) {
    R4App app; R4XStartR4Sys sys; R4XStartR4Net net; init_app(&app, &sys, &net); reply_mode = MODE_OK;
    R4Services services = r4_app_services(&app); R4ServiceConnection connection; uint32_t request = 123u, response = 0u;
    if (!r4_services_available(&services) || r4_services_open(&services, "EXAMPLE", &connection) != 0) return 1;
    if (r4_service_connection_call_struct(&connection, 7u, &request, sizeof(request), &response, sizeof(response), forever_timeout()).kind != R4_SERVICE_CALL_RESPONSE || response != request) return 2;
    if (r4_service_connection_call(&connection, 999u, 0, 0u, 0, 0u, poll_timeout()).kind != R4_SERVICE_CALL_REMOTE_FAILURE) return 3;
    if (r4_service_connection_close(&connection) != 0 || r4_service_connection_close(&connection) != R4OS_ERR_CLOSED) return 4;
    R4ServiceEndpoint endpoint; if (r4_services_register(&services, "EXAMPLE", 0u, &endpoint) != 0 || r4_service_endpoint_wait(&endpoint, poll_timeout()).kind != R4_ENDPOINT_WAIT_TIMED_OUT) return 5;
    endpoint_ready = 1; if (r4_service_endpoint_wait(&endpoint, forever_timeout()).kind != R4_ENDPOINT_WAIT_READY) return 6;
    uint32_t endpoint_payload = 0u; R4EndpointReceiveResult received = r4_service_endpoint_recv(&endpoint, &endpoint_payload, sizeof(endpoint_payload)); if (received.kind != R4_ENDPOINT_RECEIVE_MESSAGE || endpoint_payload != 42u || r4_service_endpoint_reply(&endpoint, received.header.request_id, 0, 0, 0u) != 0 || r4_service_endpoint_unregister(&endpoint) != 0) return 7;

    R4Network network = r4_app_network(&app); R4Resolver resolver = r4_network_resolver(network); R4Ipv4Address address;
    if (!r4_network_available(&network) || r4_resolver_resolve_a(&resolver, (const uint8_t *)"r4os.local", 10u, 0, forever_timeout(), &address).kind != R4_NET_RESULT_OK || address.octets[3] != 15u) return 8;
    R4SocketAddress remote = {r4_ipv4(10u, 0u, 2u, 2u), 1234u, 0u}; R4TcpSocket socket;
    if (r4_network_connect_tcp(&network, remote, forever_timeout(), &socket).kind != R4_NET_RESULT_OK || r4_tcp_socket_write(&socket, (const uint8_t *)"GET", 3u, forever_timeout()).bytes != 3u) return 9;
    uint8_t buffer[8] = {0}; if (r4_tcp_socket_read(&socket, buffer, sizeof(buffer), forever_timeout()).bytes != 2u || memcmp(buffer, "OK", 2u) != 0) return 10;
    reply_mode = MODE_WOULD_BLOCK; if (r4_tcp_socket_read(&socket, buffer, sizeof(buffer), poll_timeout()).kind != R4_NET_RESULT_WOULD_BLOCK) return 11;
    reply_mode = MODE_PEER_CLOSE; if (r4_tcp_socket_read(&socket, buffer, sizeof(buffer), forever_timeout()).kind != R4_NET_RESULT_PEER_CLOSED) return 12;
    reply_mode = MODE_OK; R4UdpSocket udp; if (r4_network_bind_udp(&network, 8000u, forever_timeout(), &udp).kind != R4_NET_RESULT_OK || r4_udp_socket_send_to(&udp, remote, (const uint8_t *)"UDP", 3u, forever_timeout()).bytes != 3u) return 13;
    R4Datagram datagram; if (r4_udp_socket_receive_from(&udp, buffer, sizeof(buffer), forever_timeout(), &datagram).kind != R4_NET_RESULT_OK || datagram.bytes != 3u) return 14;
    reply_mode = MODE_TIMEOUT; resolver = r4_network_resolver(network); if (r4_resolver_resolve_a(&resolver, (const uint8_t *)"r4os.local", 10u, 0, forever_timeout(), &address).kind != R4_NET_RESULT_TIMED_OUT) return 15;
    reply_mode = MODE_NO_SERVICE; if (r4_network_connect_tcp(&network, remote, forever_timeout(), &socket).kind != R4_NET_RESULT_NO_SERVICE || open_handles != 0u) return 16;
    return 0;
}

File: Tests/Conformance/ServiceNetworkFacadePrototype05826.zig
const std = @import("std");
const r4os = @import("r4os");

const ReplyMode = enum { ok, would_block, timeout, reset, peer_close, no_service };

var sys_table: r4os.abi.R4XStartR4Sys = .{};
var net_table: r4os.abi.R4XStartR4Net = .{};
var context: r4os.abi.R4XStartContext = .{};
var reply_mode: ReplyMode = .ok;
var endpoint_ready = false;
var open_handles: u32 = 0;
var endpoint_registered = false;
var window_reply: r4os.abi.WindowModeReply = .{};
var window_timeout = false;
var window_request: r4os.abi.WindowModeRequest = .{};

fn fakeOpen(name_z: [*:0]const u8, out: *r4os.abi.ServiceInfo) callconv(.c) i32 {
    const name = std.mem.span(name_z);
    if (reply_mode == .no_service or std.mem.eql(u8, name, "STOPPED")) return r4os.abi.service_api_result_not_running;
    const handle: u32 = if (std.mem.eql(u8, name, "DNSSVC")) 101 else if (std.mem.eql(u8, name, "TCPSVC")) 102 else if (std.mem.eql(u8, name, "UDPSVC")) 103 else 104;
    out.* = .{ .handle = handle };
    open_handles += 1;
    return r4os.abi.service_api_result_ok;
}

fn fakeClose(handle: u32) callconv(.c) i32 {
    if (handle < 101 or handle > 104 or open_handles == 0) return r4os.abi.service_api_result_bad_handle;
    open_handles -= 1;
    return r4os.abi.service_api_result_ok;
}

fn writeReply(comptime T: type, response: [*]u8, capacity: u32, value: *const T, payload: []const u8) i32 {
    const total = @sizeOf(T) + payload.len;
    if (total > capacity) return r4os.abi.service_api_result_buffer_too_small;
    @memcpy(response[0..@sizeOf(T)], std.mem.asBytes(value));
    if (payload.len != 0) @memcpy(response[@sizeOf(T)..total], payload);
    return @intCast(total);
}

fn socketLifecycle() u32 {
    return switch (reply_mode) {
        .reset => r4os.abi.net_service_socket_lifecycle_reset,
        .peer_close => r4os.abi.net_service_socket_lifecycle_peer_gone,
        .timeout => r4os.abi.net_service_socket_lifecycle_timeout,
        .would_block => r4os.abi.net_service_socket_lifecycle_would_block,
        else => r4os.abi.net_service_socket_lifecycle_active,
    };
}

fn socketStatus() u32 {
    return switch (reply_mode) {
        .timeout => r4os.abi.net_service_status_timeout,
        .would_block => r4os.abi.net_service_status_would_block,
        else => r4os.abi.net_service_status_ok,
    };
}

fn fakeCall(handle: u32, op: u16, request: [*]const u8, request_len: u32, header: *r4os.abi.ServiceMessageHeader, response: [*]u8, capacity: u32, _: u64) callconv(.c) i32 {
    header.* = .{ .op = op, .request_id = 77, .status = r4os.abi.service_api_result_ok };
    if (handle == 104) {
        if (op == r4os.abi.window_mode_op_query or op == r4os.abi.window_mode_op_request) {
            if (request_len != @sizeOf(r4os.abi.WindowModeRequest)) return r4os.abi.service_api_result_invalid;
            @memcpy(std.mem.asBytes(&window_request), request[0..request_len]);
            if (window_timeout) return r4os.abi.service_api_result_timeout;
            return writeReply(r4os.abi.WindowModeReply, response, capacity, &window_reply, "");
        }
        if (op == 7) {
            if (request_len > capacity) return r4os.abi.service_api_result_buffer_too_small;
            @memcpy(response[0..request_len], request[0..request_len]);
            return @intCast(request_len);
        }
        header.status = r4os.abi.service_api_result_invalid;
        return 0;
    }
    if (handle == 101) {
        var result: r4os.abi.NetServiceDnsResult = .{ .result = r4os.abi.dns_result_ok, .flags = r4os.abi.net_service_dns_flag_ok, .answer = .{ 10, 0, 2, 15 } };
        if (reply_mode == .timeout) result.result = r4os.abi.dns_result_timeout;
        if (reply_mode == .would_block) result.result = r4os.abi.dns_result_nxdomain;
        return writeReply(r4os.abi.NetServiceDnsResult, response, capacity, &result, "");
    }
    if (handle == 102) {
        var result: r4os.abi.NetServiceTcpResult = .{
            .action = op,
            .result = 0,
            .flags = r4os.abi.net_service_tcp_flag_ok | r4os.abi.net_service_tcp_flag_handle_valid,
            .handle = 501,
            .bytes = if (op == r4os.abi.net_service_op_tcp_write_result) @intCast(r4os.service_deadline.split(request[0..request_len]).payload.len - 4) else 0,
            .pending_rx = if (op == r4os.abi.net_service_op_tcp_poll_result) 2 else 0,
            .remote_ip = .{ 10, 0, 2, 2 },
            .remote_port = 1234,
            .lifecycle_cause = socketLifecycle(),
            .service_status = socketStatus(),
        };
        const payload = if (op == r4os.abi.net_service_op_tcp_read_result and reply_mode == .ok) "OK" else "";
        if (op == r4os.abi.net_service_op_tcp_read_result) result.bytes = @intCast(payload.len);
        if (op == r4os.abi.net_service_op_tcp_close_result or op == r4os.abi.net_service_op_tcp_close_listen_result) result.lifecycle_cause = r4os.abi.net_service_socket_lifecycle_local_close;
        return writeReply(r4os.abi.NetServiceTcpResult, response, capacity, &result, payload);
    }
    if (handle == 103) {
        var result: r4os.abi.NetServiceUdpResult = .{
            .action = op,
            .result = 0,
            .flags = r4os.abi.net_service_udp_flag_ok | r4os.abi.net_service_udp_flag_handle_valid,
            .handle = 601,
            .source_ip = .{ 10, 0, 2, 3 },
            .dest_ip = .{ 10, 0, 2, 15 },
            .source_port = 7000,
            .dest_port = 8000,
            .lifecycle_cause = socketLifecycle(),
            .service_status = socketStatus(),
        };
        const payload = if (op == r4os.abi.net_service_op_udp_recv_result and reply_mode == .ok) "UDP" else "";
        result.bytes = if (op == r4os.abi.net_service_op_udp_sendto_result) request_len - 10 else @intCast(payload.len);
        if (op == r4os.abi.net_service_op_udp_close_result) result.lifecycle_cause = r4os.abi.net_service_socket_lifecycle_local_close;
        return writeReply(r4os.abi.NetServiceUdpResult, response, capacity, &result, payload);
    }
    return r4os.abi.service_api_result_bad_handle;
}

fn fakeRegister(_: [*:0]const u8, _: u32, out: *r4os.abi.ServiceInfo) callconv(.c) i32 {
    if (endpoint_registered) return r4os.abi.service_api_result_invalid;
    endpoint_registered = true;
    out.* = .{ .handle = 201 };
    return r4os.abi.service_api_result_ok;
}

fn fakeUnregister(handle: u32) callconv(.c) i32 {
    if (!endpoint_registered or handle != 201) return r4os.abi.service_api_result_bad_handle;
    endpoint_registered = false;
    return r4os.abi.service_api_result_ok;
}

fn fakeWait(handle: u32, _: u64) callconv(.c) i32 {
    if (!endpoint_registered or handle != 201) return r4os.abi.service_api_result_bad_handle;
    return if (endpoint_ready) 1 else 0;
}

fn fakeRecv(handle: u32, header: *r4os.abi.ServiceMessageHeader, out: [*]u8, capacity: u32) callconv(.c) i32 {
    if (!endpoint_registered or handle != 201) return r4os.abi.service_api_result_bad_handle;
    if (!endpoint_ready) {
        header.* = .{};
        return 0;
    }
    const payload = [_]u32{42};
    if (capacity < @sizeOf(@TypeOf(payload))) return r4os.abi.service_api_result_buffer_too_small;
    header.* = .{ .op = 9, .request_id = 88, .payload_len = @sizeOf(@TypeOf(payload)) };
    @memcpy(out[0..@sizeOf(@TypeOf(payload))], std.mem.asBytes(&payload));
    endpoint_ready = false;
    return @sizeOf(@TypeOf(payload));
}

fn fakeReply(handle: u32, request_id: u32, status: i32, _: [*]const u8, _: u32) callconv(.c) i32 {
    return if (handle == 201 and request_id == 88 and status == 0) r4os.abi.service_api_result_ok else r4os.abi.service_api_result_invalid;
}

fn fakeTcpMarker(_: u8, _: u8, _: u8, _: u8, _: u16) callconv(.c) i32 {
    return -1;
}

fn initFacades() struct { services: r4os.Services, network: r4os.Network } {
    sys_table = .{};
    sys_table.service_open = @intFromPtr(&fakeOpen);
    sys_table.service_close = @intFromPtr(&fakeClose);
    sys_table.service_call = @intFromPtr(&fakeCall);
    sys_table.service_endpoint_register = @intFromPtr(&fakeRegister);
    sys_table.service_endpoint_unregister = @intFromPtr(&fakeUnregister);
    sys_table.service_endpoint_wait = @intFromPtr(&fakeWait);
    sys_table.service_endpoint_recv = @intFromPtr(&fakeRecv);
    sys_table.service_endpoint_reply = @intFromPtr(&fakeReply);
    net_table = .{};
    net_table.tcp_connect = @intFromPtr(&fakeTcpMarker);
    const bundle = r4os.program.Bundle{ .raw = &context, .sys = &sys_table, .net = &net_table };
    const sys = r4os.r4sys.Context.init(&bundle);
    return .{
        .services = .{ .sys = sys },
        .network = .{ .sys = sys, .net = r4os.r4net.Context.init(&bundle) },
    };
}

test "service connection and endpoint own complete lifecycles" {
    reply_mode = .ok;
    endpoint_ready = false;
    open_handles = 0;
    endpoint_registered = false;
    var facades = initFacades();
    try std.testing.expect(facades.services.available());
    var connection = switch (facades.services.open("EXAMPLE")) {
        .connection => |value| value,
        else => return error.Open,
    };
    const request: u32 = 123;
    try std.testing.expectEqual(request, switch (connection.callTyped(u32, u32, 7, &request, r4os.time_contract.timeoutForever())) {
        .value => |value| value,
        else => return error.Call,
    });
    try std.testing.expectEqual(r4os.abi.service_api_result_invalid, switch (connection.call(999, "", &.{}, r4os.time_contract.timeoutPoll())) {
        .remote_failure => |raw| raw,
        else => 0,
    });
    try std.testing.expectEqual(r4os.abi.service_api_result_ok, connection.close());
    try std.testing.expectEqual(r4os.abi.err_closed, connection.close());
    try std.testing.expect(switch (facades.services.open("STOPPED")) {
        .failure => |raw| raw == r4os.abi.service_api_result_not_running,
        else => false,
    });

    var endpoint = switch (facades.services.register("EXAMPLE", 0)) {
        .endpoint => |value| value,
        else => return error.Register,
    };
    try std.testing.expect(switch (endpoint.wait(r4os.time_contract.timeoutPoll())) {
        .timed_out => true,
        else => false,
    });
    endpoint_ready = true;
    try std.testing.expect(switch (endpoint.wait(r4os.time_contract.timeoutForever())) {
        .ready => true,
        else => false,
    });
    const received = endpoint.recvTyped([1]u32);
    const message = switch (received) {
        .message => |value| value,
        else => return error.Receive,
    };
    try std.testing.expectEqual(@as(u32, 42), message.payload[0]);
    try std.testing.expectEqual(r4os.abi.service_api_result_ok, endpoint.reply(message.header.request_id, 0, "OK"));
    try std.testing.expectEqual(r4os.abi.service_api_result_ok, endpoint.unregister());
    try std.testing.expectEqual(r4os.abi.err_closed, endpoint.unregister());
    try checkWindowModes();
}

test "resolver TCP and UDP classify lifecycle outcomes without fallback" {
    reply_mode = .ok;
    endpoint_registered = false;
    open_handles = 0;
    var facades = initFacades();
    try std.testing.expect(facades.network.available());
    var resolver = facades.network.resolver();
    const address = switch (resolver.resolveA("r4os.local", null, r4os.time_contract.timeoutForever())) {
        .address => |value| value,
        else => return error.Dns,
    };
    try std.testing.expectEqual([4]u8{ 10, 0, 2, 15 }, address.octets);
    reply_mode = .timeout;
    try std.testing.expect(switch (resolver.resolveA("r4os.local", null, r4os.time_contract.timeoutForever())) {
        .timed_out => true,
        else => false,
    });
    reply_mode = .would_block;
    try std.testing.expect(switch (resolver.resolveA("missing.local", null, r4os.time_contract.timeoutForever())) {
        .not_found => true,
        else => false,
    });

    reply_mode = .ok;
    const remote = r4os.SocketAddress{ .address = r4os.Ipv4Address.init(10, 0, 2, 2), .port = 1234 };
    var socket = switch (facades.network.connectTcp(remote, r4os.time_contract.timeoutForever())) {
        .socket => |value| value,
        else => return error.Connect,
    };
    try std.testing.expectEqual(@as(usize, 3), switch (socket.write("GET", r4os.time_contract.timeoutForever())) {
        .bytes => |bytes| bytes,
        else => 0,
    });
    var tcp_data: [8]u8 = undefined;
    try std.testing.expectEqual(@as(usize, 2), switch (socket.read(tcp_data[0..], r4os.time_contract.timeoutForever())) {
        .bytes => |bytes| bytes,
        else => 0,
    });
    try std.testing.expectEqualStrings("OK", tcp_data[0..2]);
    reply_mode = .would_block;
    try std.testing.expect(switch (socket.read(tcp_data[0..], r4os.time_contract.timeoutPoll())) {
        .would_block => true,
        else => false,
    });
    reply_mode = .timeout;
    try std.testing.expect(switch (socket.poll(r4os.time_contract.timeoutForever())) {
        .timed_out => true,
        else => false,
    });
    reply_mode = .peer_close;
    try std.testing.expect(switch (socket.read(tcp_data[0..], r4os.time_contract.timeoutForever())) {
        .peer_closed => true,
        else => false,
    });

    reply_mode = .ok;
    var listener = switch (facades.network.listenTcp(8080, r4os.time_contract.timeoutForever())) {
        .listener => |value| value,
        else => return error.Listen,
    };
    var accepted = switch (listener.accept(r4os.time_contract.timeoutForever())) {
        .socket => |value| value,
        else => return error.Accept,
    };
    reply_mode = .reset;
    try std.testing.expect(switch (accepted.poll(r4os.time_contract.timeoutForever())) {
        .reset => true,
        else => false,
    });
    reply_mode = .ok;
    try std.testing.expect(switch (listener.close(r4os.time_contract.timeoutForever())) {
        .closed => true,
        else => false,
    });

    var udp = switch (facades.network.bindUdp(8000, r4os.time_contract.timeoutForever())) {
        .socket => |value| value,
        else => return error.UdpBind,
    };
    try std.testing.expectEqual(@as(usize, 3), switch (udp.sendTo(remote, "UDP", r4os.time_contract.timeoutForever())) {
        .bytes => |bytes| bytes,
        else => 0,
    });
    var udp_data: [8]u8 = undefined;
    const datagram = switch (udp.receiveFrom(udp_data[0..], r4os.time_contract.timeoutForever())) {
        .datagram => |value| value,
        else => return error.UdpReceive,
    };
    try std.testing.expectEqual(@as(usize, 3), datagram.bytes);
    reply_mode = .would_block;
    try std.testing.expect(switch (udp.receiveFrom(udp_data[0..], r4os.time_contract.timeoutPoll())) {
        .would_block => true,
        else => false,
    });
    reply_mode = .ok;
    try std.testing.expect(switch (udp.close(r4os.time_contract.timeoutForever())) {
        .closed => true,
        else => false,
    });

    reply_mode = .no_service;
    try std.testing.expect(switch (facades.network.connectTcp(remote, r4os.time_contract.timeoutForever())) {
        .no_service => true,
        else => false,
    });
    try std.testing.expectEqual(@as(u32, 0), open_handles);
}


fn checkWindowModes() !void {
    const t = std.testing;
    reply_mode = .ok;
    window_timeout = false;
    const sys = initFacades().services.sys;
    const owner: r4os.abi.ProgramProcessHandle = .{ .instance_id = 7, .generation = 21 };
    const identity: r4os.abi.WindowModeIdentity = .{ .owner = owner, .window_id = 0, .serial = 1,
        .service = .{ .instance_id = 8, .generation = 22 }, .desktop = .{ .instance_id = 9, .generation = 23 } };
    var client: r4os.window_mode.Client = .{ .owner = owner, .window_id = 0 };
    window_reply = .{ .identity = identity };
    try t.expect(client.poll(&sys));
    window_timeout = true;
    try t.expect(!client.set(&sys, .fullscreen));
    const accepted = window_request;
    try t.expectEqualDeep(accepted, client.pending.?);
    window_timeout = false;
    window_reply = .{ .result = -3 };
    try t.expect(client.poll(&sys));
    try t.expectEqual(@as(i32, -3), client.state.result);
    try t.expect(client.pending != null);
    window_reply = .{ .identity = identity, .phase = 1, .request_id = accepted.request_id, .requested_mode = 1 };
    try t.expect(client.retry(&sys));
    try t.expectEqualDeep(accepted, window_request);
    try t.expect(client.pending != null and client.state.mode == 0);
    window_reply.phase = 2; window_reply.mode = 1;
    try t.expect(client.poll(&sys));
    try t.expect(client.pending == null and client.state.mode == 1);
    window_timeout = true;
    try t.expect(!client.set(&sys, .windowed));
    const old = client.pending.?;
    window_timeout = false;
    window_reply = .{ .identity = identity };
    window_reply.identity.service.generation += 1;
    try t.expect(client.poll(&sys));
    try t.expect(client.pending == null and client.last_error == -4);
    try t.expectEqual(@as(u32, 0), window_request.action);
    try t.expect(window_request.request_id != old.request_id);
    try t.expectEqual(@as(u32, 0), open_handles);
}
